sim/common/mop/trinkets_phase_3_52.go

```diff
@@ -534,4 +547,132 @@ func init() {
+			haste := createStatBuffAura("Re-Origination - Haste", core.ActionID{SpellID: 139121})
+			mastery := createStatBuffAura("Re-Origination - Mastery", core.ActionID{SpellID: 139120})
+
+			triggerAura := core.MakeProcTriggerAura(&character.Unit, core.ProcTrigger{
+				Name:    label,
+				Harmful: true,
+				DPM: character.NewRPPMProcManager(itemID, false, core.ProcMaskMeleeOrMeleeProc|core.ProcMaskRangedOrRangedProc, core.RPPMConfig{
+					PPM: 1.10000002384,
+				}.WithApproximateIlvlMod(1.0, 528)),
+				ICD:      duration,
+				Callback: core.CallbackOnSpellHitDealt | core.CallbackOnPeriodicDamageDealt,
+				Handler: func(sim *core.Simulation, spell *core.Spell, _ *core.SpellResult) {
+					mastery.Deactivate(sim)
+					crit.Deactivate(sim)
+					haste.Deactivate(sim)
+
+					hasMasteryRaidBuff := masteryRaidBuffs.GetActiveAura() != nil && masteryRaidBuffs.GetActiveAura().IsActive()
+					critValue := character.GetStat(stats.CritRating)
+					hasteValue := character.GetStat(stats.HasteRating)
+					// TODO: Confirm this actually excludes Mastery raid buffs
+					// At proc time, it checks how much crit, haste, and mastery you have (yes, this is a snapshot, and does NOT include the mastery raid buff).
+					// Source: http://us.battle.net/wow/en/forum/topic/7811342046?page=81#1606
+					masteryValue := character.GetStat(stats.MasteryRating) + core.TernaryFloat64(hasMasteryRaidBuff, -3000, 0)
+
+					// Determine highestStat stat with tiebreakers: Crit > Haste > Mastery
+					highestStat := stats.CritRating
+					highestValue := critValue
+					if hasteValue > highestValue {
+						highestStat = stats.HasteRating
+						highestValue = hasteValue
+					}
+					if masteryValue > highestValue {
+						highestStat = stats.MasteryRating
+						highestValue = masteryValue
+					}
+
+					switch highestStat {
+					case stats.CritRating:
+						buffStats = stats.Stats{
+							stats.CritRating:    (hasteValue + masteryValue) * 2,
+							stats.HasteRating:   -hasteValue,
+							stats.MasteryRating: -masteryValue,
+						}
+					case stats.HasteRating:
+						buffStats = stats.Stats{
+							stats.CritRating:    -critValue,
+							stats.HasteRating:   (critValue + masteryValue) * 2,
+							stats.MasteryRating: -masteryValue,
+						}
+					case stats.MasteryRating:
+						buffStats = stats.Stats{
+							stats.CritRating:    -critValue,
+							stats.HasteRating:   -hasteValue,
+							stats.MasteryRating: (critValue + hasteValue) * 2,
+						}
+					}
+
+					switch highestStat {
+					case stats.CritRating:
+						crit.Activate(sim)
+					case stats.HasteRating:
+						haste.Activate(sim)
+					case stats.MasteryRating:
+						mastery.Activate(sim)
+					}
+				},
+			})
+
+			eligibleSlots := character.ItemSwap.EligibleSlotsForItem(itemID)
+			character.AddStatProcBuff(itemID, mastery, false, eligibleSlots)
+			character.AddStatProcBuff(itemID, crit, false, eligibleSlots)
+			character.AddStatProcBuff(itemID, haste, false, eligibleSlots)
```

Comment:
This can also be cleaned up as:
```
for _, buffAura := range buffAuras {
    character.AddStatProcBuff(itemID, buffAura, false, eligibleSlots)
}
```